File: guides/micronaut-kafka/analytics/java/src/main/java/example/micronaut/AnalyticsService.java
package example.micronaut;

import jakarta.inject.Singleton;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

@Singleton
public class AnalyticsService {

    private final Map<Book, Long> bookAnalytics = new ConcurrentHashMap<>(); // <1>

    public void updateBookAnalytics(Book book) { // <2>
        bookAnalytics.compute(book, (k, v) -> {
            return v == null ? 1L : v + 1;
        });
    }

    public List<BookAnalytics> listAnalytics() { // <3>
        return bookAnalytics
                .entrySet()
                .stream()
                .map(e -> new BookAnalytics(e.getKey().getIsbn(), e.getValue()))
                .collect(Collectors.toList());
    }
}

File: guides/micronaut-kafka/analytics/kotlin/src/main/kotlin/example/micronaut/AnalyticsService.kt
package example.micronaut

import java.util.concurrent.ConcurrentHashMap
import jakarta.inject.Singleton

@Singleton
class AnalyticsService {

    private val bookAnalytics: MutableMap<Book, Long> = ConcurrentHashMap() // <1>

    fun updateBookAnalytics(book: Book) { // <2>
        bookAnalytics.compute(book) { k, v ->
            if (v == null) return@compute 1L else return@compute v + 1
        }
    }

    fun listAnalytics(): List<BookAnalytics> = // <3>
            bookAnalytics.entries.map { (key, value) -> BookAnalytics(key.isbn, value) }
}
